internal/server/v2/observation/calculation_util.go

```diff
@@ -183,29 +201,106 @@ func evalBinaryExpr(
 			}
 		}
 		if len(newOrderedFacets) > 0 {
-			if len(result.ByEntity) == 0 {
-				result.ByEntity = map[string]*pbv2.EntityObservation{}
+			result.ByEntity[entity] = &pbv2.EntityObservation{
+				OrderedFacets: newOrderedFacets,
 			}
+		}
+	}
+	return &intermediateResponse{variableObs: result}, nil
+}
+
+// Combine one VariableObservation with one constant using an operator token.
+func evalBinaryVariableConstantNodeExpr(
+	intermediate *pbv2.VariableObservation,
+	constant *float64,
+	iFirst bool, // Whether the intermediate response is the first response in the expression.
+	op token.Token,
+) (*intermediateResponse, error) {
+	result := &pbv2.VariableObservation{ByEntity: map[string]*pbv2.EntityObservation{}}
+	for entity, iEntityObs := range intermediate.ByEntity {
+		facets := iEntityObs.OrderedFacets
+		newOrderedFacets := []*pbv2.FacetObservation{}
+		for i := 0; i < len(facets); i++ {
+			newFacetId := facets[i].GetFacetId()
+			newPointStat := []*pb.PointStat{}
+			for _, obs := range facets[i].Observations {
+				iVal := obs.GetValue()
+				var val float64
+				var err error
+				if iFirst {
+					val, err = evalOp(iVal, *constant, op)
+				} else {
+					val, err = evalOp(*constant, iVal, op)
+				}
+				if err != nil {
+					return nil, err
+				}
+				newPointStat = append(newPointStat, &pb.PointStat{
+					Date:  obs.GetDate(),
+					Value: proto.Float64(val),
+				})
+			}
+			if len(newPointStat) > 0 {
+				newOrderedFacets = append(newOrderedFacets, &pbv2.FacetObservation{
+					FacetId:      newFacetId,
+					Observations: newPointStat,
+					EarliestDate: newPointStat[0].GetDate(),
+					LatestDate:   newPointStat[len(newPointStat)-1].GetDate(),
+					ObsCount:     int32(len(newPointStat)),
+				})
+			}
+		}
+		if len(newOrderedFacets) > 0 {
 			result.ByEntity[entity] = &pbv2.EntityObservation{
 				OrderedFacets: newOrderedFacets,
 			}
 		}
 	}
-	return result, nil
+	return &intermediateResponse{variableObs: result}, nil
+}
+
+// Combine two intermediateResponses using an operator token.
+func evalBinaryExpr(
+	x, y *intermediateResponse,
+	op token.Token,
+) (*intermediateResponse, error) {
+	if (x.variableObs != nil) && (y.variableObs != nil) {
+		return evalBinaryVariableObsExpr(x.variableObs, y.variableObs, op)
+	}
+	if (x.variableObs != nil) && (y.constantObs != nil) {
+		return evalBinaryVariableConstantNodeExpr(x.variableObs, y.constantObs, true /*iFirst*/, op)
+	}
+	if (x.constantObs != nil) && (y.variableObs != nil) {
+		return evalBinaryVariableConstantNodeExpr(y.variableObs, x.constantObs, false /*iFirst*/, op)
+	}
+	if (x.constantObs != nil) && (y.constantObs != nil) {
```

Comment:
I'm wondering if these 4 conditions can be abstracted into an enum - e.g. 

```
type BinaryExprType int

const (
	VarVar BinaryExprType = iota
	VarConst
	ConstVar
	ConstConst
)
```

It could potentially simplify things by not requiring 2 separate eval methods. We'd just need a trivial method to compute the enum value and then eval the intermediate responses based on it.

If it does simplify things then consider adopting that approach else ok to stay with this one.

Reply:
Sorry I was a bit confused what you mean to evaluate the intermediate response based on the enum value. I could try to combine `evalBinaryVariableObsExpr` and `evalBinaryVariableConstantNodeExpr` into one function (but they currently have different behavior). I updated some of the variable names/comments though to hopefully be more clear, but let me know if there's something in particular you'd like for this!

Reply:
Ah got it - the expressive names helped clarify things! Let's go with what you have.